### PR TITLE
PCHR-459: Fixing hrcase installation error

### PR DESCRIPTION
**Pull request by omarabuhussein**
## Problem

installing hrcase extension causing an error because the maanged entity handler was trying to create (Exiting) case type which is an already existing case type created by 1402 upgrader .

## solution

I've moved running upgraders to postInstall method where I can control weather to create or update the case type .. since the managed entity handler runs before postInstall method and 'Exiting' case type will not be there .

### CODE REVIEW

File: hrcase/CRM/HRCase/Upgrader/Base.php [hunk 1]
```diff
@@ -87,45 +93,66 @@ public function executeCustomDataFile($relativePath) {
...
     );
     return TRUE;
   }
 
   /**
-   * Run one SQL query
+   * @param string $tplFile
+   *   The SQL file path (relative to this extension's dir).
+   *   Ex: "sql/mydata.mysql.tpl".
+   * @return bool
```
Comment (MiyaNoctem) [inline hrcase/CRM/HRCase/Upgrader/Base.php]: Leave space between @params and @return.

File: hrcase/CRM/HRCase/Upgrader/Base.php [hunk 2]
```diff
@@ -219,31 +246,55 @@ public function getRevisions() {
...
   public function getCurrentRevision() {
-    // return CRM_Core_BAO_Extension::getSchemaVersion($this->extensionName);
+    $revision = CRM_Core_BAO_Extension::getSchemaVersion($this->extensionName);
+    if (!$revision) {
+      $revision = $this->getCurrentRevisionDeprecated();
+    }
+    return $revision;
+  }
+
+  private function getCurrentRevisionDeprecated() {
```
Comment (MiyaNoctem) [inline hrcase/CRM/HRCase/Upgrader/Base.php]: Missing function documentation block.

File: hrcase/hrcase.civix.php [hunk 3]
```diff
@@ -203,40 +269,97 @@ function _hrcase_civix_glob($pattern) {
...
 
 /**
- * Inserts a navigation menu item at a given place in the hierarchy
+ * Inserts a navigation menu item at a given place in the hierarchy.
  *
- * $menu - menu hierarchy
- * $path - path where insertion should happen (ie. Administer/System Settings)
- * $item - menu you need to insert (parent/child attributes will be filled for you)
- * $parentId - used internally to recurse in the menu structure
+ * @param array $menu - menu hierarchy
```
Comment (MiyaNoctem) [inline hrcase/hrcase.civix.php]: Description for each @param should go on new line?

File: hrcase/hrcase.civix.php [hunk 4]
```diff
@@ -203,40 +269,97 @@ function _hrcase_civix_glob($pattern) {
...
-  static $navId;
-
+function _hrcase_civix_insert_navigation_menu(&$menu, $path, $item) {
   // If we are done going down the path, insert menu
   if (empty($path)) {
-    if (!$navId) $navId = CRM_Core_DAO::singleValueQuery("SELECT max(id) FROM civicrm_navigation");
-    $navId ++;
-    $menu[$navId] = array (
-      'attributes' => array_merge($item, array(
+    $menu[] = array(
```
Comment (MiyaNoctem) [inline hrcase/hrcase.civix.php]: Please use short array syntax.

**Comment by omarabuhussein** [comment]
@MiyaNoctem  your comments are against auto generated files .. we don't touch auto generated files.

**Comment by omarabuhussein** [comment]
>  Question: don't all methods in hrcase.civix.php that say "Implements xxxxx hook" be marked as delegated? Or am I misunderstanding what this "Delegation" means?

not sure what civicrm really means by delegation .. but if you refer to wikipedia https://en.wikipedia.org/wiki/Delegation_(computing) 

> Delegation, in computing or computer programming, refers generally to one entity passing something to another entity.

which is kinnda happen in both hrcase.php and hrcase.civix.php .

you can see for example https://github.com/civicrm/civihr/blob/dd588bf09818cf02ee872dc185493ec43f7330f6/hrcase/hrcase.php#L31 delegate part of its work to _hrcase_civix_civicrm_postInstall which is here https://github.com/civicrm/civihr/blob/dd588bf09818cf02ee872dc185493ec43f7330f6/hrcase/hrcase.civix.php#L63.

also https://github.com/civicrm/civihr/blob/dd588bf09818cf02ee872dc185493ec43f7330f6/hrcase/hrcase.civix.php#L80

delegated its work to $upgrader->onEnable(); method .

**Pull request merged**